File: FindMinimumRotatedSortedArray.py
# -*- coding: utf-8 -*-
"""
Created on Tue Feb  4 21:13:32 2020

@author: WELCOME
"""
"""
Find Min element in rotated sorted array
Time - O(logN)
Space - O(1)
"""

class Solution:
    def findMin(self, nums: List[int]) -> int:
        low=0
        high=len(nums)-1
        while(low<=high):
            mid=(low+high)//2
            if ((mid==0 or nums[mid]<nums[mid-1]) and (mid==len(nums)-1 or nums[mid]<nums[mid+1])):
                return nums[mid]
            if nums[mid]<nums[high]:
                high=mid-1
            else:
                low=mid+1
        

File: FindMinimumRotatedSortedArray.java
class Solution {
  public int findMin(int[] nums) {

    int left = 0;
    int right = nums.length - 1;

    if (nums.length == 1)
      return nums[0];

    while (left < right) {
      int mid = left + (right - left) / 2;

      if (nums[left] <= nums[right])
        return nums[left];

      if (nums[mid] > nums[right]) {
        left = mid + 1;
      } else {
        right = mid;
      }
    }
    return nums[left];

  }
}
